straxen/url_config.py

```diff
@@ -0,0 +1,211 @@
+
+import json
+import strax
+import fsspec
+import straxen
+from numpy import isin
+import inspect
+from urllib.parse import urlparse, parse_qs
+from ast import literal_eval
+from functools import lru_cache
+
+from strax.config import OMITTED
+
+export, __all__ = strax.exporter()
+
+
+def parse_val(val):
+    try:
+        val = literal_eval(val)
+    except ValueError:
+        pass
+    return val
+
+
+@export
+class URLConfig(strax.Config):
+    """Dispatch on URL protocol.
+    unrecognized protocol returns identity
+    inspired by dasks Dispatch and fsspec fs protocols.
+    """
+    
+    _LOOKUP = {}
+    SCHEME_SEP = '://'
+    QUERY_SEP = '?'
+    PLUGIN_ATTR_PREFIX = 'plugin.'
+
+    def __init__(self, cache=0, **kwargs):
+        """
+        :param cache: number of values to keep in cache, 
+                      if set to True will cache all values
+        :param \**kwargs: additional keyword arguments accepted by strax.Option              
+        """
+        self.final_type = OMITTED
+        super().__init__(**kwargs)
+        # Ensure backwards compatibility with Option validation
+        # type of the config value can be different from the fetched value.
+        if self.type is not OMITTED:
+            self.final_type = self.type
+            self.type = OMITTED # do not enforce type on the URL
+        if cache:
+            maxsize = cache if isinstance(cache, int) else None
+            self.dispatch = lru_cache(maxsize)(self.dispatch)
+
+    @classmethod
+    def register(cls, protocol, func=None):
+        """Register dispatch of `func` on urls
+        starting with protocol name `protocol` """
+
+        def wrapper(func):
+            if isinstance(protocol, tuple):
+                for t in protocol:
+                    cls.register(t, func)
+                return func
+
+            if not isinstance(protocol, str):
+                raise ValueError('Protocol name must be a string.')
+
+            if protocol in cls._LOOKUP:
+                raise ValueError(f'Protocol with name {protocol} already registered.')
+            cls._LOOKUP[protocol] = func    
+            return func
+        return wrapper(func) if func is not None else wrapper
+
+    def dispatch(self, url, *args, **kwargs):
+        """
+        Call the corresponding method based on protocol in url.
+        chained protocols will be called with the result of the
+        previous protocol as input
+        overrides are passed to any protocol whos signature can accept them.
+        """
+        
+        # seperate the protocol name from the path
+        protocol, _, path =  url.partition(self.SCHEME_SEP)
+
+        # find the corresponding protocol method
+        meth = self._LOOKUP.get(protocol, None)
+        if meth is None:
+            # unrecongnized protocol
+            # evaluate as string-literal
+            return url
```

Comment:
Given lines 144-152 in fetch, I am wondering if this should not raise an Error. I think here you took already care of all the cases which do not follow your url logic, did not you?.